src/pytorch_lightning/loops/epoch/training_epoch_loop.py

```diff
@@ -102,7 +102,21 @@ def _is_validation_done(self) -> bool:
     @property
     def done(self) -> bool:
         """Evaluates when to leave the loop."""
-        return (self._is_training_done and self._is_validation_done) or self.trainer.should_stop
+        if self._is_training_done and self._is_validation_done:
+            return True
+
+        if self.trainer.should_stop:
+            # early stopping
+            min_epochs = self.trainer.fit_loop.min_epochs
+            should_stop_early = self.trainer.fit_loop._should_stop_early
```

Comment:
I can move this to `training_epoch_loop`and use `Trainer.min_epochs` to get that property since it is required but then it will use `fit_loop` indirectly in the epoch loop.

Reply:
can't find a way, both rely on each other for some property.